# Hacker News Post Analysis

The purpose of this project is to analyze a subset of submissions to the Hacker News website who's post were either specific questions to the Hacker News community, or post submitted related to projects products, or something interesting. I will compare the two following

- Which post recives more comments on average.
- Do posts created at a certain times receive more comments on average.

In [1]:
from csv import reader
import datetime as dt

open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
# Creating a list of just headers, and creating new dataset withouth headers
headers = hn[0]

hn = hn[1:]

print(hn[:5])



[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [5]:
'''Creating new list to filter out post that were questions,
vs post that were interesting projects/content vs others'''

ask_posts = []

show_posts = []

other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        


Ask post:  1744
show post:  1162
Other post:  17194


In [6]:
print("Ask post: ", len(ask_posts))


print("show post: ", len(show_posts))


print("Other post: ", len(other_posts))

Ask post:  1744
show post:  1162
Other post:  17194


In [14]:
''' Next will find the total number of comments for both ask
and show post to determine which post recived more comments
avg'''



def avg_comments(post_name):
    total_ask_comments = 0
    for post in post_name:
        comments = int(post[4])
        total_ask_comments += comments
    avg_ask_comments = total_ask_comments / len(post_name)
    return (avg_ask_comments)

print("avg ask comments: ", avg_comments(ask_posts))
print("avg show comments: ", avg_comments(show_posts))
    

avg ask comments:  14.038417431192661
avg show comments:  10.31669535283993


Analysis show that on average,Ask Post recieve 14.04 comments per post, while Show post recived 10.32 comments per post. Showing that Ask post are a slightly more popular.  

I will now focus the remaining analysis on ask post to identify if ask posts created at a certain time more likely to attack comments.

In [19]:
# change date field to datetime object to make parsing easier.

d1_format = "%m/%d/%Y %H:%M"
results_list = []

for row in ask_posts:
    date = row[-1]
    comments= int(row[4])
    new_date = dt.datetime.strptime(date,d1_format)
    results_list.append([new_date, comments])

print("example of results list: \n", results_list[:5])

example of results list: 
 [[datetime.datetime(2016, 8, 16, 9, 55), 6], [datetime.datetime(2015, 11, 22, 13, 43), 29], [datetime.datetime(2016, 5, 2, 10, 14), 1], [datetime.datetime(2016, 8, 2, 14, 20), 3], [datetime.datetime(2015, 10, 15, 16, 38), 17]]


In [26]:
counts_by_hour = {}
comments_by_hour = {}


for row in results_list: 
    hour = row[0].strftime('%H')
    comments = row[1]
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
        
        
print(counts_by_hour)

print(comments_by_hour)

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [27]:
''' Below shows the number of ask posts created during each
hour of the day. Then a second frequency table that contains 
the number of comments asks posts created at each hour received'''

        
print(counts_by_hour)

print('\n')

print(comments_by_hour)


{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [28]:
''' Next I will create a list of a list to show the average 
number of comments per posts created during each hour of the 
day'''

avg_by_hour = []

for item in comments_by_hour:
    avg_by_hour.append([item,comments_by_hour[item]/counts_by_hour[item]])
    
    
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [44]:
#final sorting

swap_avg_by_hour = []

for item in avg_by_hour:
    swap_avg_by_hour.append([item[1],item[0]])

print(swap_avg_by_hour)

print('\n')

sorted_swap = sorted(swap_avg_by_hour,reverse=True)

print("top 5 Hours for Ask Posts Comments \n", sorted_swap[:5])
print('\n')
print("top 5 Hours for Ask Posts Comments - Formatted \n")
for item in sorted_swap[:5]:
    time = int(item[1])
    avg = item[0]
    date_obj = dt.time(hour=time)
    f_time= date_obj.strftime("%H:%S")
    
    
    print("{}:{:,.2f} average comments per post".format(f_time,avg))
    


[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


top 5 Hours for Ask Posts Comments 
 [[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


top 5 Hours for Ask Posts Comments - Formatted 

15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average com

# Final Analysis

Based on my , the top five hours that a post will acquire the most comments is the 5 PM hour with the average number of comments are 38.59. I believe this could be attributed to the average time user's are off work.

top 5 Hours for Ask Posts Comments - Formatted 

15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post